File: airesources/CSharp/hlt/Entity.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

/// <summary>
/// 
/// </summary>
public enum DockingStatus { undocked, docking, docked, undocking }

/// <summary>
/// 
/// </summary>
public enum EntityType { Planet, Ship };

/// <summary>
/// Entity Information
/// </summary>
public class EntityInfo {

 private int owner;
 private int id;
 private EntityType type;

 /// <summary>
 /// Constructor for EntityInfo
 /// </summary>
 /// <param name="owner">Owner of Entity</param>
 /// <param name="id">Id of Entity</param>
 /// <param name="type">Type of Entity(Planet, Ship)</param>
 public EntityInfo (int owner, int id, EntityType type) {
 this.owner = owner;
 this.id = id;
 this.type = type;
    }

    /// <summary>
    /// Owner of the Entity (Ship or Planet)
    /// </summary>
    /// <returns></returns>
    public int Owner {
        get {
            return owner;
        }
    }

    /// <summary>
    /// Id of the Entity (Ship or Planet)
    /// </summary>
    /// <returns></returns>
    public int Id {
        get {
            return id;
        }
    }

    /// <summary>
    /// Type of Entity (Ship or Planet)
    /// </summary>
    /// <returns>EntityType</returns>
    public EntityType EntityType {
        get {
            return type;
        }
    }

    public override string ToString () {
        StringBuilder s = new StringBuilder ();
        s.AppendLine ("");
        s.AppendLine ("Entity Info");
        s.AppendLine ("Type-" + this.EntityType);
        s.AppendLine ("Id- " + this.Id);
        s.AppendLine ("Owner- " + this.Owner);
        s.AppendLine ("");
        return s.ToString ();
    }
}

/// <summary>
/// Base class for entities like Planet and Ship
/// </summary>
public class Entity {
    EntityInfo entityInfo;
    Position position;

    double radius;

    /// <summary>
    /// Id of an entity
    /// </summary>
    /// <returns>Id of an entity</returns>
    public EntityInfo EntityInfo {
        get {
            return entityInfo;
        }
        protected set {
            entityInfo = value;
        }
    }

    /// <summary>
    /// Position of an entity
    /// </summary>
    /// <returns>Position of an entity</returns>
    public Position Position {
        get {
            return position;
        }
        protected set {
            position = value;
        }

    }

    /// <summary>
    /// Radius of the entity
    /// </summary>
    /// <returns>Radius of an entity</returns>
    public double Radius {
        get {
            return radius;
        }
        protected set {
            radius = value;
        }

    }

    /// <summary>
    /// Gets euclidien distance between the entity and target
    /// </summary>
    /// <param name="target">Position in space or an entity</param>
    /// <returns>Distance from entity to given target</returns>
    public double GetDistance (Position target) {
        var dx = this.position.X - target.X;
        var dy = this.position.Y - target.Y;
        return Math.Sqrt (dx * dx + dy * dy);
    }

    /// <summary>
    /// Get angle between entity and target
    /// </summary>
    /// <param name="target">Position in space or an entity</param>
    /// <returns>Angle between entity snd given target</returns>
    public double GetAngle (Position target) {
        return Math.Atan2 (target.Y - this.position.Y, target.X - this.position.X) * 180.0 / Math.PI;
    }

    /// <summary>
    /// 
    /// </summary>
    /// <param name="target"></param>
    /// <param name="minimumDistance"></param>
    /// <returns></returns>
    public Position GetClosestPointToEntity (Entity target, double minimumDistance = 3) {
        var radius = target.Radius + minimumDistance;
        var angle = target.GetAngle (this.Position);
        return new Position (target.Position.X + radius * Math.Cos (angle), target.Position.Y + radius * Math.Sin (angle));
    }

    public override string ToString () {
        StringBuilder s = new StringBuilder ();
        s.AppendLine ("");
        s.AppendLine ("Entity Info");
        s.AppendLine ("Owner-" + entityInfo.Owner);
        s.AppendLine ("Id- " + this.entityInfo.Id);
        s.AppendLine ("Position- " + this.Position.ToString ());
        s.AppendLine ("Radius- " + this.Radius);
        s.AppendLine ("");
        return s.ToString ();
    }
}

/// <summary>
/// 
/// </summary>
public class Player {

    public List<Ship> Ships {
        get {
            return ships;
        }

        set {
            ships = value;
        }
    }

    public int Id {
        get {
            return id;
        }
    }

    private List<Ship> ships;
    private int id;

    public Player (int id) {
        this.id = id;
        ships = new List<Ship> ();
    }
}

/// <summary>
/// 
/// </summary>
public class Planet : Entity {

    /// <summary>
    /// 
    /// </summary>
    /// <returns></returns>
    public int DockingSpots {
        get {
            return dockingSpots;
        }
    }

    /// <summary>
    /// 
    /// </summary>
    /// <returns></returns>
    public int CurrentProduction {
        get {
            return currentProduction;
        }
    }

    /// <summary>
    /// 
    /// </summary>
    /// <returns></returns>
    public int RemainingProduction {
        get {
            return remainingProduction;
        }
    }

    /// <summary>
    /// 
    /// </summary>
    /// <returns></returns>
    public int Health {
        get {
            return health;
        }
    }

    /// <summary>
    /// 
    /// </summary>
    /// <returns></returns>
    public List<int> DockedShips {
        get {
            return dockedShips;
        }
    }

    int dockingSpots;
    int currentProduction;
    int remainingProduction;
    int health;
    List<int> dockedShips;

    /// <summary>
    /// 
    /// </summary>
    /// <param name="tokens"></param>
    /// <returns></returns>
    public Planet (Queue<string> tokens) : base () {
        int id = int.Parse (tokens.Dequeue ());
        this.Position = new Position (Double.Parse (tokens.Dequeue ()), Double.Parse (tokens.Dequeue ()));
        this.health = int.Parse (tokens.Dequeue ());
        this.Radius = Double.Parse (tokens.Dequeue ());
        this.dockingSpots = int.Parse (tokens.Dequeue ());
        this.currentProduction = int.Parse (tokens.Dequeue ());
        this.remainingProduction = int.Parse (tokens.Dequeue ());
        int owner = -1;
        if (int.Parse (tokens.Dequeue ()) == 1) {
            owner = int.Parse (tokens.Dequeue ());
        } else {
            tokens.Dequeue ();
        }

        this.EntityInfo = new EntityInfo (owner, id, EntityType.Planet);
        int dockedShips = int.Parse (tokens.Dequeue ());
        this.dockedShips = new List<int> (dockedShips);
        for (int i = 0; i < dockedShips; i++) {
            this.dockedShips.Add (int.Parse (tokens.Dequeue ()));
        }

        Log.Information (this.ToString (), LogingLevel.Game);
    }

    /// <summary>
    /// 
    /// </summary>
    /// <returns></returns>
    public bool isOwned () {
        return this.EntityInfo.Owner >= 0 ? true : false;
    }

    /// <summary>
    /// 
    /// </summary>
    /// <param name="id"></param>
    /// <returns></returns>
    public bool isFull (int id) {
        return this.dockedShips.Count >= this.dockingSpots ? true : false;
    }

    /// <summary>
    /// 
    /// </summary>
    /// <returns></returns>
    public override string ToString () {
        StringBuilder s = new StringBuilder ();
        s.AppendLine ("");
        s.AppendLine ("Planet Info");
        if (this.EntityInfo.Owner != -1) {
            s.AppendLine ("Owner-" + this.EntityInfo.Owner);
        } else {
            s.AppendLine ("Owner-None");
        }

        s.AppendLine ("Id- " + this.EntityInfo.Id);
        s.AppendLine ("Position- " + this.Position.ToString ());
        s.AppendLine ("Radius- " + this.Radius);
        s.AppendLine ("Docking Spots- " + this.dockingSpots);
        s.AppendLine ("Health- " + this.health);
        s.AppendLine ("Current Production- " + this.currentProduction);
        s.AppendLine ("");
        return s.ToString ();
    }
}

/// <summary>
/// Ship class
/// </summary>
public class Ship : Entity {

    Velocity velocity;
    int health;
    DockingStatus dockingStatus = DockingStatus.docked;
    int dockedPlanet;
    int dockingProgress;
    int weaponCooldown;

    /// <summary>
    /// Magnitude of the ships velocity
    /// </summary>
    /// <returns></returns>
    public Velocity Velocity {
        get {
            return velocity;
        }
    }

    /// <summary>
    /// Health of the ship
    /// </summary>
    /// <returns></returns>
    public int Health {
        get {
            return health;
        }
    }

    /// <summary>
    /// Docking status for the ship
    /// </summary>
    /// <returns></returns>
    public DockingStatus DockingStatus {
        get {
            return dockingStatus;
        }
    }

    /// <summary>
    /// If the ship is docked, then what planet
    /// </summary>
    /// <returns></returns>
    public int DockedPlanet {
        get {
            return dockedPlanet;
        }
    }

    /// <summary>
    /// Docking takes a few turns, this tells you how close you are to completing it
    /// </summary>
    /// <returns></returns>
    public int DockingProgress {
        get {
            return dockingProgress;
        }
    }

    /// <summary>
    /// Constructor for the Ship Class
    /// </summary>
    /// <param name="owner">Owner of this ship</param>
    /// <param name="tokens">Tokens from the game environment to pazrse</param>
    public Ship (int owner, Queue<String> tokens) : base () {
        this.EntityInfo = new EntityInfo (owner, int.Parse (tokens.Dequeue ()), EntityType.Ship);
        this.Position = new Position (Double.Parse (tokens.Dequeue ()), Double.Parse (tokens.Dequeue ()));
        this.health = int.Parse (tokens.Dequeue ());
        this.velocity = new Velocity (Double.Parse (tokens.Dequeue ()), Double.Parse (tokens.Dequeue ()));
        this.dockingStatus = (DockingStatus) Enum.Parse (typeof (DockingStatus), tokens.Dequeue ());
        this.dockedPlanet = int.Parse (tokens.Dequeue ());
        this.dockingProgress = int.Parse (tokens.Dequeue ());
        this.weaponCooldown = int.Parse (tokens.Dequeue ());
        this.Radius = Constants.ShipRadius;
        Log.Information (this.ToString (), LogingLevel.Game);
    }

    /// <summary>
    /// Checks if the ship is close enough to the planet to dock
    /// </summary>
    /// <param name="planet">The planet to check the docking status for</param>
    /// <returns>True if the planet can be docked to, else false</returns>
    public bool CanDock (Planet planet) {
        if (this.GetDistance (planet.Position) <= planet.Radius + Constants.DockRadius) {
            Log.Information ("Ship: " + this.EntityInfo.Id + " Can Dock: True ", LogingLevel.User);
            return true;
        }

        Log.Information ("Ship: " + this.EntityInfo.Id + " Can Dock: False ", LogingLevel.User);
        return false;
    }

    /// <summary>
    /// Give a speed an angle, move the ship
    /// </summary>
    /// <param name="speed">Speed to apply</param>
    /// <param name="angle">Angle to apply the speed on</param>
    /// <returns>Move command</returns>
    public string Move (double speed, int angle) {
        return String.Format ("t {0} {1} {2}", this.EntityInfo.Id, speed, angle);
    }

    /// <summary>
    /// Dock the ship to the planet
    /// </summary>
    /// <param name="planet">Planet to dock to</param>
    /// <returns>Dock command</returns>
    public string Dock (Planet planet) {
        return String.Format ("d {0} {1}", this.EntityInfo.Id, planet.EntityInfo.Id);
    }

    /// <summary>
    /// Undock from the planet
    /// </summary>
    /// <param name="planet">The planet to undock from</param>
    /// <returns>Undock command</returns>
    public string UnDock (Planet planet) {
        return String.Format ("u {0}", this.EntityInfo.Id);
    }

    /// <summary>
    /// Given a targert, navigate the ship with basic pathfinding
    /// </summary>
    /// <param name="target">Target to navigate too</param>
    /// <param name="map">Game map</param>
    /// <param name="speed">Speed at which to move</param>
    /// <param name="avoidObstacles">True, if you want to avoid obstacles</param>
    /// <param name="maxCorrection">Maximum number if course corrections to factor into the move</param>
    /// <returns></returns>
    public string Navigate (Position target, GameMap map, Double speed, bool avoidObstacles = true, int maxCorrection = 90) {
        if (maxCorrection <= 0)
            return string.Empty;
        var distance = this.GetDistance (target);
        var angle = this.GetAngle (target);
        var obstacles = map.GetObstaclesBetween (this, target);
        if (avoidObstacles == true && obstacles.Count > 0) {
            Log.Information ("Avoiding Obstacles", LogingLevel.User);
            var newTargetDX = Math.Cos (angle + 1 * distance);
            var newTargetDY = Math.Sin (angle + 1 * distance);
            var newTarget = new Position (this.Position.X + newTargetDX, this.Position.Y + newTargetDY);
            return this.Navigate (newTarget, map, speed, true, maxCorrection - 1);
        }

        speed = distance >= speed ? speed : distance;
        return this.Move (Convert.ToInt32 (speed), Convert.ToInt32 (angle));
    }

    /// <summary>
    /// Pretty printed list of ship properties
    /// </summary>
    /// <returns>string representation of the ships properties</returns>
    public override string ToString () {
        StringBuilder s = new StringBuilder ();
        s.AppendLine ("");
        s.AppendLine ("Ship Info");
        s.AppendLine ("Owner-" + this.EntityInfo.Owner);
        s.AppendLine ("Id- " + this.EntityInfo.Id);
        s.AppendLine ("Position- " + this.Position.ToString ());
        s.AppendLine ("Health- " + this.health);
        s.AppendLine ("Velocity- " + this.velocity.ToString ());
        s.AppendLine ("DockingStatus- " + this.dockingStatus);
        s.AppendLine ("Docking Planet- " + this.dockedPlanet);
        s.AppendLine ("Docking Progress- " + this.dockingProgress);
        s.AppendLine ("");
        return s.ToString ();
    }

    /// <summary>
    /// Parse game tokens into ships
    /// </summary>
    /// <param name="owner">Owner of this ship</param>
    /// <param name="tokens">Set of tokens from the game</param>
    /// <returns>List of ships</returns>
    public static List<Ship> Parse (int owner, Queue<String> tokens) {
        int numberOfShips = int.Parse (tokens.Dequeue ());
        List<Ship> ships = new List<Ship> ();
        for (int i = 0; i < numberOfShips; i++) {
            ships.Add (new Ship (owner, tokens));
        }
        return ships;
    }
}

/// <summary>
/// 
/// </summary>
public class Position {

    private double xPos;
    private double yPos;

    /// <summary>
    /// 
    /// </summary>
    /// <param name="xPos"></param>
    /// <param name="yPos"></param>
    public Position (double xPos, double yPos) {
        this.xPos = xPos;
        this.yPos = yPos;
    }

    /// <summary>
    /// 
    /// </summary>
    /// <returns></returns>
    public double X {
        get {
            return xPos;
        }
    }

    /// <summary>
    /// 
    /// </summary>
    /// <returns></returns>
    public double Y {
        get {
            return yPos;
        }
    }

    /// <summary>
    /// 
    /// </summary>
    /// <param name="target"></param>
    /// <returns></returns>
    public double GetDistance (Position target) {
        var dx = this.X - target.X;
        var dy = this.Y - target.Y;
        return Math.Sqrt (dx * dx + dy * dy);
    }

    /// <summary>
    /// 
    /// </summary>
    /// <param name="o"></param>
    /// <returns></returns>
    public override bool Equals (Object o) {
        if (this == o)
            return true;
        if (o == null || this.GetType () != o.GetType ())
            return false;

        Position position = (Position) o;

        if (position.X != xPos)
            return false;

        return position.Y == yPos ? true : false;
    }

    /// <summary>
    /// 
    /// </summary>
    /// <returns></returns>
    public override int GetHashCode () {
        return (int) xPos ^ (int) yPos;
    }

    /// <summary>
    /// 
    /// </summary>
    /// <returns></returns>
    public override String ToString () {
        return "Position{" +
            "xPos=" + xPos +
            ", yPos=" + yPos +
            '}';
    }
}

/// <summary>
/// 
/// </summary>
public class Velocity {

    private double x;
    private double y;

    /// <summary>
    /// 
    /// </summary>
    /// <param name="xVelocity"></param>
    /// <param name="yVelocity"></param>
    public Velocity (double xVelocity, double yVelocity) {
        this.x = X;
        this.y = Y;
    }

    /// <summary>
    /// 
    /// </summary>
    /// <returns></returns>
    public double X {
        get {
            return x;
        }
    }

    /// <summary>
    /// 
    /// </summary>
    /// <returns></returns>
    public double Y {
        get {
            return y;
        }
    }

    /// <summary>
    /// Get magnitude of velocity
    /// </summary>
    /// <returns>Magnitude of velocity as string</returns>
    public double GetMagnitude () {
        return Math.Sqrt (Math.Pow (x, 2) + Math.Pow (y, 2));
    }

    /// <summary>
    /// Get angular direction from velocity
    /// </summary>
    /// <returns>Angular direction as double</returns>
    public double GetAngle () {
        return Math.Atan2 (y, x);
    }

    /// <summary>
    /// String representation of velocity
    /// </summary>
    /// <returns>String representation of velocity</returns>
    public override String ToString () {
        return "Velocity{" +
            "xVelocity=" + x +
            ", yVelocity=" + y +
            '}';
    }
}

/// <summary>
/// 
/// </summary>
public class Size {

    private int width;
    private int height;

    /// <summary>
    /// 
    /// </summary>
    /// <param name="xVelocity"></param>
    /// <param name="yVelocity"></param>
    public Size (int width, int height) {
        this.width = width;
        this.height = height;
    }

    /// <summary>
    /// 
    /// </summary>
    /// <returns></returns>
    public int Width {
        get {
            return width;
        }
    }

    /// <summary>
    /// 
    /// </summary>
    /// <returns></returns>
    public int Height {
        get {
            return height;
        }
    }

    /// <summary>
    /// String representation of velocity
    /// </summary>
    /// <returns>String representation of velocity</returns>
    public override String ToString () {
        return "Map Size{" +
            "Width=" + width +
            ", Height=" + height +
            '}';
    }
}

File: airesources/Rust/src/hlt/entity.rs


use std::cmp::min;

use hlt::parse::Decodable;
use hlt::command::Command;
use hlt::constants::{DOCK_RADIUS, SHIP_RADIUS, MAX_SPEED};
use hlt::player::Player;
use hlt::game_map::GameMap;

#[derive(PartialEq, Debug, Clone, Copy)]
pub struct Position(pub f64, pub f64);

impl Decodable for Position {
    fn parse<'a, I>(tokens: &mut I) -> Position
    where
        I: Iterator<Item = &'a str>,
    {

        let x = f64::parse(tokens);
        let y = f64::parse(tokens);
        return Position(x, y);
    }
}

#[derive(PartialEq, Debug)]
pub enum DockingStatus {
    UNDOCKED = 0,
    DOCKING = 1,
    DOCKED = 2,
    UNDOCKING = 3,
}

impl Decodable for DockingStatus {
    fn parse<'a, I>(tokens: &mut I) -> DockingStatus
    where
        I: Iterator<Item = &'a str>,
    {

        let i = i32::parse(tokens);
        return match i {
            0 => DockingStatus::UNDOCKED,
            1 => DockingStatus::DOCKING,
            2 => DockingStatus::DOCKED,
            3 => DockingStatus::UNDOCKING,
            _ => panic!(format!("Not a valid docking status: {:?}", i)),
        };
    }
}

#[derive(PartialEq, Debug)]
pub struct Ship {
    pub id: i32,
    pub position: Position,
    pub hp: i32,
    pub velocity_x: f64,
    pub velocity_y: f64,
    pub docking_status: DockingStatus,
    pub docked_planet: Option<i32>,
    pub progress: i32,
    pub cooldown: i32,
}

impl Ship {
    pub fn thrust(&self, magnitude: i32, angle: i32) -> Command {
        Command::Thrust(self.id, magnitude, angle)
    }

    pub fn dock(&self, planet: &Planet) -> Command {
        Command::Dock(self.id, planet.id)
    }

    #[allow(dead_code)]
    pub fn undock(&self) -> Command {
        Command::Undock(self.id)
    }

    pub fn can_dock(&self, planet: &Planet) -> bool {
        self.calculate_distance_between(planet) <= (DOCK_RADIUS + planet.radius)
    }

    pub fn navigate<T: Entity>(&self, target: &T, game_map: &GameMap, max_corrections: i32) -> Option<Command> {
        if max_corrections <= 0 {
            return None
        }
        let angular_step = 1.0;
        let speed = MAX_SPEED;
        let distance = self.calculate_distance_between(target);
        let angle = self.calculate_angle_between(target);
        if game_map.obstacles_between(self, target) {
            let new_target_dx = f64::cos((angle + angular_step).to_radians()) * distance;
            let new_target_dy = f64::sin((angle + angular_step).to_radians()) * distance;
            let Position(self_x, self_y) = self.position;
            let new_target = Position(self_x + new_target_dx, self_y + new_target_dy);
            self.navigate(&new_target, game_map, max_corrections - 1)
        } else {
            Some(self.thrust(min(speed, distance as i32), angle as i32))
        }
    }
}

impl Decodable for Ship {
    fn parse<'a, I>(tokens: &mut I) -> Ship
    where
        I: Iterator<Item = &'a str>,
    {

        let id = i32::parse(tokens);
        let position = Position::parse(tokens);
        let hp = i32::parse(tokens);
        let velocity_x = f64::parse(tokens);
        let velocity_y = f64::parse(tokens);
        let docking_status = DockingStatus::parse(tokens);
        let docked_planet_raw = i32::parse(tokens);
        let docked_planet = match docking_status {
            DockingStatus::UNDOCKED => None,
            _ => Some(docked_planet_raw),
        };
        let progress = i32::parse(tokens);
        let cooldown = i32::parse(tokens);

        return Ship {
            id,
            position,
            hp,
            velocity_x,
            velocity_y,
            docking_status,
            docked_planet,
            progress,
            cooldown,
        };
    }
}

#[derive(PartialEq, Debug)]
pub struct Planet {
    pub id: i32,
    pub position: Position,
    pub hp: i32,
    pub radius: f64,
    pub num_docking_spots: i32,
    pub current_production: i32,
    pub remaining_resources: i32,
    pub owner: Option<i32>,
    pub docked_ships: Vec<i32>,
}

impl Planet {
    pub fn is_owned(&self) -> bool {
        self.owner.is_some()
    }
}

impl Decodable for Planet {
    fn parse<'a, I>(tokens: &mut I) -> Planet
    where
        I: Iterator<Item = &'a str>,
    {

        let id = i32::parse(tokens);
        let position = Position::parse(tokens);
        let hp = i32::parse(tokens);
        let radius = f64::parse(tokens);
        let num_docking_spots = i32::parse(tokens);
        let current_production = i32::parse(tokens);
        let remaining_resources = i32::parse(tokens);
        let owner = Option::parse(tokens);
        let docked_ships = Vec::parse(tokens);

        return Planet {
            id,
            position,
            hp,
            radius,
            num_docking_spots,
            current_production,
            remaining_resources,
            owner,
            docked_ships,
        };
    }
}

#[derive(PartialEq, Debug)]
pub struct GameState {
    pub players: Vec<Player>,
    pub planets: Vec<Planet>,
}

impl Decodable for GameState {
    fn parse<'a, I>(tokens: &mut I) -> GameState
    where
        I: Iterator<Item = &'a str>,
    {

        let players = Vec::parse(tokens);
        let planets = Vec::parse(tokens);

        return GameState { players, planets };
    }
}

pub trait Entity : Sized {
    fn get_position(&self) -> Position;
    fn get_radius(&self) -> f64;

    fn calculate_distance_between<T: Entity>(&self, target: &T) -> f64 {
        let Position(x1, y1) = self.get_position();
        let Position(x2, y2) = target.get_position();
        f64::sqrt((x2-x1).powi(2) + (y2-y1).powi(2))
    }

    fn calculate_angle_between<T: Entity>(&self, target: &T) -> f64 {
        let Position(x1, y1) = self.get_position();
        let Position(x2, y2) = target.get_position();
        (f64::atan2(y2-y1, x2-x1).to_degrees() + 360.0) % 360.0
    }

    fn closest_point_to<T: Entity>(&self, target: &T, min_distance: f64) -> Position {
        let angle = target.calculate_angle_between(self);
        let radius = target.get_radius() + min_distance;
        let Position(target_x, target_y) = target.get_position();
        let x = target_x + radius * f64::cos(angle.to_radians());
        let y = target_y + radius * f64::sin(angle.to_radians());

        Position(x, y)
    }
}

impl Entity for Ship {
    fn get_position(&self) -> Position {
        self.position
    }

    fn get_radius(&self) -> f64 {
        SHIP_RADIUS
    }
}

impl Entity for Planet {
    fn get_position(&self) -> Position {
        self.position
    }

    fn get_radius(&self) -> f64 {
        self.radius
    }
}

impl Entity for Position {
    fn get_position(&self) -> Position {
        *self
    }

    fn get_radius(&self) -> f64 {
        0.0
    }
}